lib/publiccloud/utils.pm

```diff
@@ -79,15 +79,36 @@ sub register_addon {
     record_info('SUSEConnect time', 'The command SUSEConnect -r ' . get_addon_fullname($addon) . ' took ' . (time() - $cmd_time) . ' seconds.');
 }
 
+sub deregister_addon {
+    my ($remote, $addon) = @_;
+    my $arch = get_var('PUBLIC_CLOUD_ARCH', 'x86_64');
+    $arch = "aarch64" if ($arch eq "arm64");
+    my $timestamp = utc_timestamp();
+    record_info($addon, "Going to deregister '$addon' addon\nUTC: $timestamp");
+    my $cmd_time = time();
+    # ssh_remove_suseconnect_product($remote, $name, $version, $arch, $params, $timeout, $retries, $delay)
+    my ($timeout, $retries, $delay) = (300, 3, 120);
+    if ($addon =~ /ltss/) {
+        ssh_remove_suseconnect_product($remote, get_addon_fullname($addon), '${VERSION_ID}', $arch, "-r " . get_required_var('SCC_REGCODE_LTSS'), $timeout, $retries, $delay);
+    } elsif (is_sle('<15') && $addon =~ /tcm|wsm|contm|asmm|pcm/) {
+        ssh_remove_suseconnect_product($remote, get_addon_fullname($addon), '`echo ${VERSION} | cut -d- -f1`', $arch, '', $timeout, $retries, $delay);
+    } elsif (is_sle('<15') && $addon =~ /sdk|we/) {
+        ssh_remove_suseconnect_product($remote, get_addon_fullname($addon), '${VERSION_ID}', $arch, '', $timeout, $retries, $delay);
+    } else {
+        ssh_remove_suseconnect_product($remote, get_addon_fullname($addon), undef, $arch, '', $timeout, $retries, $delay);
+    }
+    record_info('SUSEConnect time', 'The command SUSEConnect -d ' . get_addon_fullname($addon) . ' took ' . (time() - $cmd_time) . ' seconds.');
+}
+
 sub registercloudguest {
     my ($instance) = @_;
-    my $regcode = get_required_var('SCC_REGCODE');
+    my $regcode = (is_byos()) ? '-r ' . get_required_var('SCC_REGCODE') : '';
     my $remote = $instance->username . '@' . $instance->public_ip;
-    # not all images currently have registercloudguest pre-installed .
+    # not all BYOS images currently have registercloudguest pre-installed .
     # in such a case,we need to regsiter against SCC and install registercloudguest with all needed dependencies and then
     # unregister and re-register with registercloudguest
-    if ($instance->run_ssh_command(cmd => "sudo which registercloudguest > /dev/null; echo \"registercloudguest\$?\" ", proceed_on_failure => 1) =~ m/registercloudguest1/) {
-        $instance->retry_ssh_command(cmd => "sudo SUSEConnect -r $regcode", timeout => 420, retry => 3, delay => 120);
+    if (is_byos() && $instance->run_ssh_command(cmd => "sudo which registercloudguest > /dev/null; echo \"registercloudguest\$?\" ", proceed_on_failure => 1) =~ m/registercloudguest1/) {
```

Comment:
if you check current master all calls of `registercloudguest` function come with condition `if(is_byos())` or `if(!is_on_demand())` . while generally I like idea to "move" this `if` into function and don't have it "outside" . I would call this cleaner approach . But currently I see two problems : 
1.  Current PR is not "refactoring of registercloudguest function" . and the fact that you mix this change with addition of new test make it harder to review for both the test and the refactoring. It would be just easier to tackle both if they come split. And I am speaking not just two different commits but really to separate PRs 
2. While you introduce `is_byos`  internally in function you haven't drop duplicated checks which we have on every call of this function. I would say it is overkill to do this check two times .

Reply:
Regarding 1st point there is one this which could fully drop it - necessity of having them in one PR . If I overlook it please point me to it ...

Reply:
The `registercloudguest` is now as it was before. Good point, thank you